### ISSUE TITLE
Tensorflow 2.0 AdaDelta Optimizer

### ISSUE THREAD

**Issue by vivekvekariya** [tensorflow/tensorflow#32950]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug, a feature request, or a significant problem with documentation (for small docs fixes please send a PR instead).
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
- **TensorFlow installed from (source or binary)**:
- **TensorFlow version (use command below)**:
- **Python version**:
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with:

```bash
python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"
```

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Comment by vivekvekariya** [comment]
I was using Keras to train my neural network, With Loss function as "Mean Square Error" and optimizer as "AdaDelta". Neural Network trained on MNIST dataset for 5 epochs was giving test accuracy of 92.5%.

Using same set of layers, loss function, optimizer and initialization methods, my neural network model in tensorflow 2.0 is giving output of just 15%.
With other optimizers in TF2 gradient descent is working proper giving test accuracy of more than 90% on MNIST dataset.

I am wondering why is the problem only with AdaDelta? Am I missing some seeds or parameters to be feed to optimizer?

**Comment by RaphaelMeudec** [comment]
@vivekvekariya Can you provide a snippet for both your Keras & TF2.0 training? Would be easier to debug.

**Comment by vivekvekariya** [comment]
keras.initializers.RandomNormal(mean=0.0, stddev=0.05, seed=1)
<img width="520" alt="tf2 Implementation" src="https://user-images.githubusercontent.com/38262106/66058292-68c33c00-e53a-11e9-8c08-79cca0d32bcb.png">
<img width="539" alt="Keras Implementation" src="https://user-images.githubusercontent.com/38262106/66058293-695bd280-e53a-11e9-92f0-30da3b317e91.png">

**Comment by gowthamkpr** [comment]
@vivekvekariya Can you please provide a github gist of this issue. Thanks!

**Comment by vivekvekariya** [comment]
@gowthamkpr 
Kindly find link to my github gist on this issue: 
https://gist.github.com/vivekvekariya/19d0b5a1af104d6cbac7425aa85bf408

**Comment by gowthamkpr** [comment]
@vivek-rane Thanks for your input. I tried running mnist example against adadelta and adam and yes 'adadelta' has much worse performance compared to adam (as you can see [here](https://colab.sandbox.google.com/gist/gowthamkpr/bcb1ead49a720ac75fbc38522dca6106/beginner.ipynb)) and this is explained [here](https://www.reddit.com/r/MachineLearning/comments/3y84hr/how_does_adam_compare_to_adadelta/).

**Comment by Nephalen** [comment]
I'm getting a similar issue. I'm training a model to approximate Q value in Reinforcement Learning using AdamOptimizer.

In tensorflow 1.14, TD error starts from 9.0 and continues to drop.

However, in tensorflow 2.0, TD error starts from 9.0 but continues to increase and stabilizes around 25 for some reasons.

**Comment by vivek-rane** [comment]
@gowthamkpr I think you meant to tag @vivekvekariya

**Comment by Nephalen** [comment]
> 
> 
> I was using Keras to train my neural network, With Loss function as "Mean Square Error" and optimizer as "AdaDelta". Neural Network trained on MNIST dataset for 5 epochs was giving test accuracy of 92.5%.
> 
> Using same set of layers, loss function, optimizer and initialization methods, my neural network model in tensorflow 2.0 is giving output of just 15%.
> With other optimizers in TF2 gradient descent is working proper giving test accuracy of more than 90% on MNIST dataset.
> 
> I am wondering why is the problem only with AdaDelta? Am I missing some seeds or parameters to be feed to optimizer?

May I ask if you tried optimizers like tf.keras.optimizers.Adam and tf.compat.v1.train.AdamOptimizer? I have a project showing different result using tf.compat.v1.train.AdamOptimizer.

**Comment by gowthamkpr** [comment]
@Nephalen Can you please try it with different data sets like flowers, cifar and see if the issue still persists. I have tried it on Mnist and I didn't notice any error as mentioned above.

**Comment by Nephalen** [comment]
@gowthamkpr I've tested on another supervised data set and it was similar.

The problem is, the project I'm working on is not supervised learning, as I mentioned above. For that reinforcement learning model, the training processes between TF <=1.14 and TF >=1.15 are vastly different.

There is another post stating a similar issue but I'm not sure whether its poster tested on TF 1.15 yet. [link](https://github.com/tensorflow/tensorflow/issues/33003#issue-501769956)

I couldn't find any documentation mentioning any update to the optimizers. My guess is whatever the reason behind this is not intended.

**Comment by Nephalen** [comment]
@gowthamkpr I think I figure out the reason. [link](https://github.com/tensorflow/tensorflow/issues/33471#issue-508537917)

**Comment by gowthamkpr** [comment]
As mentioned in issue #33471, Yes in TF 1.14 it shows 0 but in TF 1.15rc3 and TF 2.0 its showing 100. I think this is clearly a bug

Here are my github gists [Tf 1.14](https://colab.sandbox.google.com/gist/gowthamkpr/1fd2d4e745b093af63325bfbd7dbefb7/untitled198.ipynb), [TF 1.15rc3](https://colab.sandbox.google.com/gist/gowthamkpr/edf243d59c16cbe5709754d8d346636e/untitled199.ipynb) and [TF2.0](https://colab.sandbox.google.com/gist/gowthamkpr/0fefab16eefe321e944f7fe528376765/untitled197.ipynb)

**Comment by tanzhenyu** [comment]
We changed the default learning rate in 2.0. Can you double check and see if you're using the same learning rate?

**Comment by Nephalen** [comment]
@tanzhenyu I'm terribly sorry for the confusion. What I meant above was I figured out the cause of higher loss in my project and created another issue. It seems to be the result of incorrect weight update. [#33471](https://github.com/tensorflow/tensorflow/issues/33471#issue-508537917)

And just for the record, I explicitly kept learning rate constant across experiments.

**Comment by tanzhenyu** [comment]
@Nephalen I see. It looks like a different issue and resolved then.
@vivekvekariya I think this is due to different default learning rate, can you explicitly change the code to:
opt = tf.keras.optimizers.Adadelta(learning_rate=1.0, ...)
model.compile(opt, ...)

Closing it for now. Let us know if any other issues come up.
